### PR TITLE
[fix] [METEOR-1030] Prevent additional image groups in metadata being created for multi-channel z-stacks

### PR DESCRIPTION
**Pull request by patrickcleeve2**
When saving z-stacks additional images (groups) were being created in the metadata. This is because the check assumed all data was 5D (CTZYX) or 2D (YX) and only checked shape[0] assuming it was the channel dim, for if the group could be merged.

Since the z-stacks are 3D (ZYX) they were being added as a separate group, with incorrect channel information (TheC, FirstC) which caused problems with downstream readers (bio-formats).

When images were re-imported, they are reshaped into 5D, so when re-exporting them, the image was 'fixed'. This only occurred for multi-channel z-stacks, as the other checks caught other cases.

### CODE REVIEW

File: src/odemis/dataio/tiff.py [hunk 1]
```diff
@@ -1221,13 +1224,20 @@ def _findImageGroups(das):
             or prev_da.metadata.get(model.MD_ROTATION, 0) != da.metadata.get(model.MD_ROTATION, 0)
             or prev_da.metadata.get(model.MD_SHEAR, 0) != da.metadata.get(model.MD_SHEAR, 0)
            ):
-            # new group
-            group_ifd = current_ifd
+            # don't add a new group if passed group check and it's z-stacks
+            if prev_da.metadata.get(model.MD_DIMS, None) == "ZYX" and dims == "ZYX" and da.ndim == 3 and prev_da.ndim == 3:
```
Comment (pieleric) [inline src/odemis/dataio/tiff.py:1228]: I think it's overly catching cases. Essentially, it's now merging any DataArray which is of shape ZYX. Including possibly SEM data, or ZYX of different sizes, or ZYX at different positions in XY.

I would suggest to just modify the original condition to be more specific and say "if previous image and current image both have C dimension == 1", which the default C dimension being of length 1. 
You can use something like this to read the C dimension:
```
        dims = da.metadata.get(model.MD_DIMS, "CTZYX"[-da.ndim::])
        len_c = 1
        if "C" in dims:
            len_c = da.shape[dims.index("C")]
```

File: src/odemis/dataio/tiff.py [hunk 2]
```diff
@@ -1209,8 +1209,11 @@ def _findImageGroups(das):
     for da in das:
         # check if it can be part of the current group (compare just to the previous DA)
         dims = da.metadata.get(model.MD_DIMS, "CTZYX"[-da.ndim::])
-        if (prev_da is None
```
Comment (pieleric) [inline src/odemis/dataio/tiff.py:1212]: I feel that this code is extra difficult to read because of the inverted logic. Maybe you could inverted back, with something like:
```
if prev_da is not None and prev_da.shape == da.shape...:
    log("merging")
else:
    new group
```

**Review by pieleric — CHANGES_REQUESTED** [review]
Please don't forget to add a test case. For this type of corner case, it's super important, otherwise, in a couple of years, we break it again without noticing.

File: src/odemis/dataio/tiff.py [hunk 3]
```diff
@@ -1206,28 +1206,55 @@ def _findImageGroups(das):
...
+        # if there is no previous data array, create a new group
+        # if there is more than one channel, create a new group
+        # if the dimensions are different, create a new group
+        # if WL_IN or WL_OUT are not present (looks like FM image) (the metadata is missing), create a new group
+        # if the metadata is different, create a new group
+        if prev_da is None:
+            # create the first image group
+            group_ifd = current_ifd
+            logging.debug(f"creating initial image group: group id: {group_ifd}, current_ifd: {current_ifd}")
+
```
Comment (pieleric) [inline src/odemis/dataio/tiff.py:1242]: I don't think it's necessary to have a special `if` here. It just does the same as for the `elif`. => merge into a single `if` statement.

File: src/odemis/dataio/tiff.py [hunk 4]
```diff
@@ -1206,28 +1206,55 @@ def _findImageGroups(das):
...
-            or (model.MD_IN_WL not in da.metadata or model.MD_OUT_WL not in da.metadata)
+
+
+        plen_c, len_c = 1, 1
+        if "C" in dims:
+            len_c = da.shape[dims.index("C")]
+        if prev_da is not None:
+            prev_dims = prev_da.metadata.get(model.MD_DIMS, "CTZYX"[-prev_da.ndim::])
+            if "C" in prev_dims:
+                plen_c = prev_da.shape[prev_dims.index("C")]
```
Comment (pieleric) [inline src/odemis/dataio/tiff.py:1230]: Instead of explicitly computing it, you can do the same as `prev_da`: start with `plen_c = None`, and at the end of the loop body, do a `plen_c = len_c`.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Comment by K4rishma** [comment]
It would be a worthwhile to check if the test cases involving findImageGroups by using .export are happy. The relevant test  files are img_test.py, stream_test.py, calibration_test.py, dataio_test.py, stiff_test.py, and tiif_test.py.

**Review by K4rishma — APPROVED** [review]

File: src/odemis/dataio/tiff.py [hunk 5]
```diff
@@ -1206,28 +1206,55 @@ def _findImageGroups(das):
...
+            # create a new group
             group_ifd = current_ifd
+            logging.debug(f"creating new image group: group id: {group_ifd}, current_ifd: {current_ifd}")
         groups.setdefault(group_ifd, []).append(da)
+        logging.debug(f"end of data array: group id: {group_ifd}, current_ifd: {current_ifd}, current groups: {groups.keys()}")
 
         # increase ifd by the number of planes
         current_ifd += _countNeededIFDs(da)
         prev_da = da
+    logging.debug(f"found {len(groups)} groups of images. final groups: {groups.keys()}")
```
[the remark this answers is not in the document]
Comment (patrickcleeve2) [reply src/odemis/dataio/tiff.py:1257]: remove most, leaving one final message to assist future debugging

File: src/odemis/dataio/tiff.py (hunk 2 above)
Comment (patrickcleeve2) [reply src/odemis/dataio/tiff.py:1212]: I've inverted it, and update comments. should be easier to read

File: src/odemis/dataio/tiff.py [hunk 6]
```diff
@@ -1206,28 +1206,50 @@ def _findImageGroups(das):
...
-            or prev_da.metadata.get(model.MD_HW_NAME, None) != da.metadata.get(model.MD_HW_NAME, None)
-            or prev_da.metadata.get(model.MD_HW_VERSION, None) != da.metadata.get(model.MD_HW_VERSION, None)
-            or prev_da.metadata.get(model.MD_PIXEL_SIZE) != da.metadata.get(model.MD_PIXEL_SIZE)
-            or prev_da.metadata.get(model.MD_POS) != da.metadata.get(model.MD_POS)
-            or prev_da.metadata.get(model.MD_ROTATION, 0) != da.metadata.get(model.MD_ROTATION, 0)
-            or prev_da.metadata.get(model.MD_SHEAR, 0) != da.metadata.get(model.MD_SHEAR, 0)
-           ):
-            # new group
+
+        plen_c, len_c = 1, 1
```
Comment (pieleric) [inline src/odemis/dataio/tiff.py:1224]: You shouldn't reset `plen_c` to 1 every time.
Just set it to `None` before the `for` loop, like `prev_da`.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Review by pieleric — APPROVED** [review]
